internal/consumer.go

```diff
@@ -234,7 +235,9 @@ func (consumer *StreamConsumer) HandleTxnError(producer sarama.AsyncProducer, me
 
 func (consumer *StreamConsumer) handleTxnError(producer sarama.AsyncProducer, message *sarama.ConsumerMessage, session sarama.ConsumerGroupSession, err error, defaulthandler func() error) {
 	fmt.Printf("Message consumer: unable to process transaction: %+v", err)
-	for {
+	retryCount := 0
+	maxRetries := 30
+	for retryCount < maxRetries {
```

Comment:
A transaction **must be** canceled if some error occurs, for exactly-once. So this is good for infinite loop, but this doesn't look good for business policy. In my opinion, it would be better to delay processing (time.Sleep) or log if cross the attempt threshold in a loop. (actually  my suggestion is another business feature)

Reply:
If a transaction error occurs, it will one day be processed if the producer is okay, so no worry to infinite loop. If there is a problem with the producer, it is prevented by handling `producer.TxnStatus()& sarama.ProducerTxnFlagFatalError`